mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+
+        # Get reward
+        reward = next_timestep.reward
+        meaned_reward = jax.tree_util.tree_map(
+            lambda x: jnp.repeat(x[:, jnp.newaxis, ...], cfg.system.num_agents, axis=1),
+            jnp.mean(next_timestep.reward, axis=-1),
+        )
+
+        reward = jax.lax.select(
+            (jnp.zeros_like(reward) + cfg.system.mean_reward).astype(int), meaned_reward, reward
+        )
+
+        transition = Transition(obs, action, reward, done)
+        transition = jax.tree_util.tree_map(
+            lambda x: x[:, jnp.newaxis, ...], transition
+        )  # Add dummy time dim
+        next_buffer_state = rb.add(buffer_state, transition)
+
+        # Next obs and done for learner state
+        next_obs = next_timestep.observation
+        next_done = next_timestep.last()[
+            ..., jnp.newaxis
+        ]  # make compatible with network input and transition storage in next step
+
+        # Repack
+        new_interact_state = InteractionState(
+            next_action_selection_state, next_env_state, next_buffer_state, next_obs, next_done
+        )
+
+        return new_interact_state, next_timestep.extras["episode_metrics"]
+
+    def switch_leading_axes(arr: chex.Array) -> chex.Array:
+        """Switches the first two axes, generally used for BT -> TB."""
+        arr: Dict[str, chex.Array] = jax.tree_map(lambda x: jax.numpy.swapaxes(x, 0, 1), arr)
+        return arr
+
+    # TRAIN LEVEL 3 and 4
+    def prep_inputs_to_scannedrnn(obs: Observation, done: chex.Array) -> chex.Array:
+        """
+        Prepares the inputs to the RNN network for either
+        getting q values or the eps-greedy distribution.
+        """
+
+        hidden_state = ScannedRNN.initialize_carry(
+            (cfg.system.sample_batch_size, obs.agents_view.shape[2]), cfg.network.hidden_state_dim
+        )
+        obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+        done = switch_leading_axes(done)  # (B, T) -> (T, B)
+        obs_done = (obs, done)
+
+        return hidden_state, obs_done
+
+    # TRAIN LEVEL 3
+    def q_loss_fn(
+        q_online_params: FrozenVariableDict, obs: Array, done: Array, action: Array, target: Array
+    ) -> Tuple[Array, Metrics]:
+        """The portion of the calculation to grad, namely online apply and mse with target."""
+
+        hidden_state, obs_done = prep_inputs_to_scannedrnn(
+            obs, done
+        )  # axes switched here to scan over time
+
+        _, q_online = q_net.apply(
+            q_online_params, hidden_state, obs_done, method="get_q_values"
+        )  # get online q values of all actions
+        q_online = switch_leading_axes(q_online)  # (T, B, ...) -> (B, T, ...)
+        q_online = jnp.squeeze(
+            jnp.take_along_axis(q_online, action[..., jnp.newaxis], axis=-1), axis=-1
+        )  # get the q values of the taken actions and remove extra dim
+        q_loss = jnp.mean((q_online - target) ** 2)  # mse
```

Comment:
Up to you, I think this reads really nice, mean -> square -> error
```suggestion
        q_loss = jnp.mean(jnp.square(q_online - target))  # mse
```

Reply:
I agree :)